src/dbnode/storage/namespace/schema_registry.go

```diff
@@ -0,0 +1,177 @@
+				zap.Stringer("namespace", id))
+		}
+		return nil
+	}
+
+	if _, ok := history.GetLatest(); !ok {
+		return fmt.Errorf("can not set empty schema history for %v", id.String())
+	}
+
+	sr.Lock()
+	defer sr.Unlock()
+
+	// TODO [haijun] use generated map for optimized map lookup.
+	current, ok := sr.registry[id.String()]
+	if ok {
+		if !history.Extends(current.Get().(SchemaHistory)) {
+			return fmt.Errorf("can not update schema registry to one that does not extends the existing one")
+		}
+	} else {
+		sr.registry[id.String()] = xwatch.NewWatchable()
+	}
+
+	sr.registry[id.String()].Update(history)
+	return nil
+}
+
+func (sr *schemaRegistry) GetLatestSchema(id ident.ID) (SchemaDescr, error) {
+	if !sr.protoEnabled {
+		return nil, nil
+	}
+
+	history, err := sr.getSchemaHistory(id)
+	if err != nil {
+		return nil, err
+	}
+	schema, ok := history.GetLatest()
+	if !ok {
+		return nil, fmt.Errorf("schema history is empty for namespace %v", id.String())
+	}
+	return schema, nil
+}
+
+func (sr *schemaRegistry) GetSchema(id ident.ID, schemaId string) (SchemaDescr, error) {
+	if !sr.protoEnabled {
+		return nil, nil
+	}
+
+	history, err := sr.getSchemaHistory(id)
+	if err != nil {
+		return nil, err
+	}
+	schema, ok := history.Get(schemaId)
+	if !ok {
+		return nil, fmt.Errorf("schema of version %v is not found for namespace %v", schemaId, id.String())
+	}
+	return schema, nil
+}
+
+func (sr *schemaRegistry) getSchemaHistory(id ident.ID) (SchemaHistory, error) {
+	sr.RLock()
+	defer sr.RUnlock()
+
+	history, ok := sr.registry[id.String()]
+	if !ok {
+		return nil, fmt.Errorf("schema history is not found for %v", id.String())
+	}
+	return history.Get().(SchemaHistory), nil
+}
+
+
+func (sr *schemaRegistry) RegisterListener(
+	nsID ident.ID,
+	listener SchemaListener,
+) (xclose.SimpleCloser, error) {
+	if !sr.protoEnabled {
+		return nil, nil
+	}
+
+	sr.RLock()
+	defer sr.RUnlock()
+
+	watchable, ok := sr.registry[nsID.String()]
+	if !ok {
+		return nil, fmt.Errorf("schema not found for namespace: %v", nsID.String())
+	}
+
+	_, watch, _ := watchable.Watch()
+
+	// We always initialize the watchable so always read
+	// the first notification value
+	<-watch.C()
+
+	// Deliver the current schema
+	if schema, ok := watchable.Get().(SchemaHistory).GetLatest(); ok {
+		listener.SetSchema(schema)
+	} else if sr.logger != nil {
+		sr.logger.Error("can not update schema listener with empty schema history")
+	}
+
+	// Spawn a new goroutine that will terminate when the
+	// watchable terminates on the close of the runtime options manager
+	go func() {
+		for range watch.C() {
+			if schema, ok := watchable.Get().(SchemaHistory).GetLatest(); ok {
+				listener.SetSchema(schema)
+			} else if sr.logger != nil {
+				sr.logger.Error("can not update schema listener with empty schema history")
+			}
+		}
+	}()
+
+	return watch, nil
+}
+
+func (sr *schemaRegistry) Close() {
+	sr.RLock()
+	defer sr.RUnlock()
```

Comment:
Seems like this should be `.Lock()` and `.Unlock()` as its modifying state no?

Reply:
good catch